### ISSUE TITLE
Feature: Conditional Bindings

### ISSUE THREAD

**Issue by mutech** [mutech/aka-ios-beacon#2]
## Rationale
- Simplify binding expressions already supporting conditionals (f.e. table view data source binding/cell mappings)
- Make view models even more presentation independent (and thus more flexible and reusable).
## Syntax

Introduce a new syntax for conditional bindings:
### Example

```
$when(<TypeA>) 
    "CellForAs"
$when(<TypeB>)  
    <CustomUITableViewCellType> { cellIdentifier: "CellForB" }
$when("SELF someCondition:argument == YES" { argument: 1.2 })
    "CellForSomeConditionSatisfied"
$else
    "DefaultCell"
```

Currently, cell mappings look like this:

```
[ { predicate: <TypeA>,
    cellIdentifier: "CellForAs" },
  { predicate: <TypeB>,
    cellIdentifier: "CellForBs", type: <CustomUITableViewCellType> },
  { predicate: "SELF someCondition:argument == YES" { argument: 1.2 },
    cellIdentifier: "CellForSomeConditionSatisfied" },
  { cellIdentifier: "DefaultCell" }
]
```
### Grammar

```
<conditional> ::= <whenClause> + <elseClause> ?
<whenClause> ::= ('$when' | '$whenNot') '(' <predicateBindingExpression> ')' <bindingExpression>
<elseClause> ::= '$else' <bindingExpression>
```

Optionally, the key words '$inspect' (or '$switch', '$case', ...) and '$end' could be added. $inspect could be used to change the data context for the when clause and $end would make it possible to nest conditionals (not sure if that should be supported though). Both could be optional.

```
<conditional> ::= <inspectClause> ? <whenClause> + <elseClause> ? '$end' ?
<whenClause> ::= '$when' '(' <predicateBindingExpression> ')' <bindingExpression>
<elseClause> ::= '$else' <bindingExpression>
<inspectClause> ::= '$inspect' '(' <keyPathBindingExpression> ')'
```
### Validation Rules
- Conditional binding expressions can validly replace binding expressions of all other types as long as all of their  result binding expressions are valid according to the surrounding binding type.
### Observation

Conditional bindings observe changes that can lead to predicates changing the value (due to predicate configuration or data context changes).

They however only have to observe leading predicates up until the first one that evaluates to true. Succeeding predicates can be ignored unless the active (satisfied) predicate changes its value to false.
### Resulting Binding Activation

When the conditional binding starts observing changes, it iterates over when clauses and evaluates their predicates until one is satisfied. The first predicate that is satisfied (by the data context) is selected and the result binding expression is used to create a binding.

If the data context used in when clauses changes or non constant predicates change (f.e. an NSPredicate parameter bound to a key path), the conditional binding re-evaluated invalidated predicates.

When relevant predicates change, a previously used result binding will stopObservingChanges (thus be released) and the new result binding will be created and startObservingChanges.

**Comment by mutech** [comment]
Experimental support for this feature is implemented in a248af3a6d2054fc5b3f978bef8dd6962b6c7e92

**Comment by mutech** [comment]
Implemented in v0.3
